web-ui/src/utils/index.ts

```diff
@@ -63,3 +65,58 @@ export const PLACEHOLDER_VALUES = {
   pipeline_name: 'Price Checker',
   pipeline_description: 'Analyze e-commerce prices'
 }
+
+// A way to throw errors in async code so they can be catched with ErrorBoundary
+// in react. Pretty silly, but it works.
+//
+// See:
+// https://medium.com/trabe/catching-asynchronous-errors-in-react-using-error-boundaries-5e8a5fd7b971
+export const useAsyncError = () => {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const [_, setError] = useState()
+  return useCallback(
+    (e: any) => {
+      setError(() => {
+        throw e
+      })
+    },
+    [setError]
+  )
+}
+
+// Read from a stream, yelding one line at a time.
+//
+// Adapted from:
+// https://developer.mozilla.org/en-US/docs/Web/API/ReadableStreamDefaultReader/read#example_2_-_handling_text_line_by_line
+export async function* readLineFromStream(response: Response) {
+  const utf8Decoder = new TextDecoder('utf-8')
+  if (!response.body) {
+    throw new Error('No body when fetching request.')
+  }
+  const reader = response.body.getReader()
+  let { value: chunk, done: readerDone } = await reader.read()
+  let decodedChunk = chunk ? utf8Decoder.decode(chunk, { stream: true }) : ''
+
+  const re = /\r\n|\n|\r/gm
+  let startIndex = 0
+
+  for (;;) {
+    const result = re.exec(decodedChunk)
+    if (!result) {
+      if (readerDone) {
+        break
+      }
+      const remainder = decodedChunk.substring(startIndex)
+      ;({ value: chunk, done: readerDone } = await reader.read())
```

Comment:
what's with the semicolon here?

Reply:
No clue tbh -- added by formatter... originally from here https://developer.mozilla.org/en-US/docs/Web/API/ReadableStreamDefaultReader/read#example_2_-_handling_text_line_by_line

where it's just 
`({ value: chunk, done: readerDone } = await reader.read());`
But if I do that the formatter moves it up to the end of the previous line.

If I do instead,
`const { value: chunk, done: readerDone } = await reader.read()`

it complains that `readerDone` is used before the declaration. (I guess it doesn't realize it's declared first a few lines above the for loop.

The function is terrible, but since it's copy-paste from the documentation so I expect it to work.